### PR TITLE
Strided masked reductions: sum, amax. Testing of masked reductions.

### PR DESCRIPTION
**Pull request by pearu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66784
* #66783
* #66386
* #66385
* __->__ #65990

cc @nikitaved @pearu @cpuhrsch

Differential Revision: [D31729532](https://our.internmc.facebook.com/intern/diff/D31729532)

### CODE REVIEW

**Comment by IvanYashchuk** [comment]
Was it agreed somewhere to put these new functions in `torch.sparse` module as `torch.sparse.masked_*`?
`torch.sparse` is for tensors in sparse format. I think it's better to keep the Python prototype of masked reductions separate from the sparse module.

**Comment by pearu** [comment]
> Was it agreed somewhere to put these new functions in `torch.sparse` module as `torch.sparse.masked_*`? 

Yes, it was: https://docs.google.com/document/d/1dcgnk9ZIJpv81vL_2ZqTWTJOrdsnDX2XSQVTu99Jbtk/edit?disco=AAAAOQ-v_WM

`torch.sparse` is for tensors in sparse format. I think it's better to keep the Python prototype of masked reductions separate from the sparse module.

I agree. However, the plan was to change the location later.

**Comment by IvanYashchuk** [comment]
How much work is it to create `torch.masked` module and mark it as a prototype in documentation? Adding to just `torch.*` is also better and marking the individual functions as a prototype. Just because sparse module's documentation says it's in beta status doesn't make it a default choice for non-related functionality. The "later" might not happen ever and would make the sparse module to be even in worse shape than it is now.

**Comment by pearu** [comment]
> Was it agreed somewhere to put these new functions in `torch.sparse` module as `torch.sparse.masked_*`?

The new agreement (from 2021-10-04 torch.sparse meet) is to use `torch._masked` module and drop `masked_` function prefices. For example, `torch._masked.sum` is the function to compute the masked sum.

**Review by albanD — COMMENTED** [review]
Looks mostly ok. Small questions

**Comment by pearu** [comment]
@albanD re autograd support for masked reductions: all autograd tests (in torch_ops.py) pass with `torch._masked.sum` but not for `torch._masked.amax`. In particular, for elements with output mask equal to `False`, the `_get_numerical_vJu` produces `nan` values because it operates with the reduction operation identity values that for `amax` is `-inf`. When changing the `amax` identity value to `torch.finfo(dtype).min`, the autograd tests pass.

So, this raises a question if we should actually use `torch.finfo(dtype).min` as an identity value for `amax`?
Or, we could use `torch.finfo(dtype).min` when `input.requires_grad` is `True`, otherwise use `-inf`, would that be too confusing to users?

**Review by albanD — COMMENTED** [review]
Looks good. Only small details.
I'll let Ivan do a final review as well.

**Review by albanD — APPROVED** [review]
Thanks!

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31729532).
